File: src/py_pkg/py_pkg/subscriber.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class MinimalSubscriber(Node):

    def __init__(self):
        super().__init__("minimal_subscriber")
        self.subscription = self.create_subscription(
            String, "topic", self.callback, 10
        )  # 创建订阅者，订阅 topic，队列大小为 10

    def callback(self, message):
        self.get_logger().info(f"Received: '{message.data}'")


def main(args=None):
    rclpy.init(args=args)  # 初始化 ROS 2 上下文
    minimal_subscriber = MinimalSubscriber()  # 创建订阅者节点实例
    rclpy.spin(minimal_subscriber)  # 运行节点（阻塞式）
    minimal_subscriber.destroy_node()  # 销毁节点
    rclpy.shutdown()  # 关闭 ROS 2


if __name__ == "__main__":
    main()

File: src/cpp_pkg/src/subscriber.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class MinimalSubscriber : public rclcpp::Node { // 继承 rclcpp::Node 类，表明这是一个 C++ ROS2 节点
  public:
    MinimalSubscriber() : Node("minimal_subscriber") {
        // 创建 Subscription 对象，监听 topic 话题，队列大小为 10
        subscription_ = this->create_subscription<std_msgs::msg::String>(
            "topic", 10, std::bind(&MinimalSubscriber::callback, this, std::placeholders::_1));
    }

  private:
    void callback(const std_msgs::msg::String::SharedPtr message) {
        RCLCPP_INFO(this->get_logger(), "Received: '%s'", message->data.c_str()); // 打印接收到的消息
    }
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);                            // 初始化 ROS2
    rclcpp::spin(std::make_shared<MinimalSubscriber>()); // 运行节点
    rclcpp::shutdown();                                  // 关闭 ROS2
    return 0;
}